governance/interface.go

```diff
@@ -0,0 +1,103 @@
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the klaytn library. If not, see <http://www.gnu.org/licenses/>.
+
+package governance
+
+import (
+	"github.com/klaytn/klaytn/blockchain/types"
+	"github.com/klaytn/klaytn/common"
+	"github.com/klaytn/klaytn/consensus/istanbul"
+	"github.com/klaytn/klaytn/params"
+	"github.com/klaytn/klaytn/storage/database"
+)
+
+type Engine interface {
+	HeaderEngine
+}
+
+type HeaderEngine interface {
+	// Cast votes from API
+	AddVote(key string, val interface{}) bool
+	ValidateVote(vote *GovernanceVote) (*GovernanceVote, bool)
+
+	// Access database for voting states
+	CanWriteGovernanceState(num uint64) bool
+	WriteGovernanceState(num uint64, isCheckpoint bool) error
+
+	// Access database for network params
+	ReadGovernance(num uint64) (uint64, map[string]interface{}, error)
+	WriteGovernance(num uint64, data GovernanceSet, delta GovernanceSet) error
+
+	// Compose header.Vote and header.Governance
+	GetEncodedVote(addr common.Address, number uint64) []byte
+	GetGovernanceChange() map[string]interface{}
+
+	// Intake header.Vote and header.Governance
+	VerifyGovernance(received []byte) error
+	ClearVotes(num uint64)
+	WriteGovernanceForNextEpoch(number uint64, governance []byte)
+	UpdateCurrentSet(num uint64)
+	HandleGovernanceVote(
+		valset istanbul.ValidatorSet, votes []GovernanceVote, tally []GovernanceTallyItem,
+		header *types.Header, proposer common.Address, self common.Address) (
+		istanbul.ValidatorSet, []GovernanceVote, []GovernanceTallyItem)
+
+	// Get internal fields
+	ChainId() uint64
+	InitialChainConfig() *params.ChainConfig
+	GetVoteMap() map[string]VoteStatus
+	GetGovernanceTallies() []GovernanceTallyItem
+	PendingChanges() map[string]interface{}
+	Votes() []GovernanceVote
+	IdxCache() []uint64
+	IdxCacheFromDb() []uint64
+
+	NodeAddress() common.Address
+	TotalVotingPower() uint64
+	MyVotingPower() uint64
+	BlockChain() blockChain
+	DB() database.DBManager
+
+	// Set internal fields
+	SetNodeAddress(addr common.Address)
+	SetTotalVotingPower(t uint64)
+	SetMyVotingPower(t uint64)
+	SetBlockchain(chain blockChain)
+	SetTxPool(txpool txPool)
+
+	// Get network params
+	GovernanceMode() string
+	GoverningNode() common.Address
+	UnitPrice() uint64
+	CommitteeSize() uint64
+	Epoch() uint64
+	ProposerPolicy() uint64
+	DeferredTxFee() bool
+	MinimumStake() string
+	MintingAmount() string
+	ProposerUpdateInterval() uint64
+	Ratio() string
+	StakingUpdateInterval() uint64
+	UseGiniCoeff() bool
+	GetGovernanceValue(key int) interface{}
+	GetGovernanceItemAtNumber(num uint64, key string) (interface{}, error)
+	GetItemAtNumberByIntKey(num uint64, key int) (interface{}, error)
+	GetGoverningInfoAtNumber(num uint64) (bool, common.Address, error)
+	GetMinimumStakingAtNumber(num uint64) (uint64, error)
+
+	// Test harness
+	AddGovernanceCacheForTest(num uint64, config *params.ChainConfig)
```

Comment:
- Why public?
  - This method is used in two places.
    - tests/hard_fork_test.go:90
    - tests/klay_test_blockchain_test.go:126
  - Because the tests are in other package, it has to be public.

- Why method?
  - Originally there was a global function `AddGovernanceCacheForTest(gov, num, config)`, but I replaced it by a method. I thought [[public method](https://github.com/klaytn/klaytn/pull/1297/files#diff-933e859a7cd2b63def11f6ba992e6348d986aac783f5dd00a49d16c4e46c73b7R754)] is more natural than [[global function that calls argument's private method](https://github.com/klaytn/klaytn/pull/1297/files#diff-933e859a7cd2b63def11f6ba992e6348d986aac783f5dd00a49d16c4e46c73b7L1076)]. 

```diff
- governance.AddGovernanceCacheForTest(gov, num, config)
+ gov.AddGovernanceCacheForTest(num, config)
```

- Why test harness?
  - The governance package holds a few in-memory caches (idxCache, itemCache). The cache is may not properly setup under certain testing circumstances. Unfortunately, some code may assume that the cache is correctly set. That is why testing code directly modifies the internal cache. I do not want to fix this part now.

Reply:
Let's concentrate only on the issue of whether the test-purpose function defined in the core interface is appropriate or not.

If that function must be public, the global function seems not bad in my view because that function is test-purpose. The private function call seems not to matter too in the test purpose. Moreover, it seems not private because the function is defined in the same package. 

If you think your change is better, then you may keep your way.

Reply:
I agree that the test-purpose function need not be in a public interface.
Fixed in https://github.com/klaytn/klaytn/pull/1297/commits/35d39d8112f5dfe89ae097e43529689d52977a92. Reverted to old way.